### PR TITLE
Support PostgreSQL INTERVAL, TIMETZ, BYTEA, and BIT types in Live Migration and expand IT coverage

### PR DESCRIPTION
**Pull request by jsuhani-2026**
Adds Datastream-to-Spanner live migration support for PostgreSQL `INTERVAL`, `TIMETZ`, `BYTEA`, `BIT` and `VARBIT`  string data types, and expands end-to-end integration test coverage for already supported PostgreSQL data types.

**Data Type Support & Behavior Changes:**
1. INTERVAL
- Previous Behavior: The values were completely unhandled during JSON extraction, resulting in them being silently migrated to Spanner as null.
- New Behavior: The pipeline now correctly processes the interval Avro record emitted by Datastream (which contains months, hours, and micros fields). It computes the correct period and duration, formatting the result into a standard ISO-8601 duration string (e.g., P1Y2M or PT5H0.123456S), which successfully maps to Spanner STRING columns.

2. TIMETZ
- Previous Behavior: The values were completely unhandled during JSON extraction, resulting in them being silently migrated to Spanner as null.
- New Behavior: The pipeline now correctly processes the timeTz Avro record emitted by Datastream (which contains time in microseconds and a timezone offset). It converts these fields into a standard ISO offset time string (e.g., 23:59:59+10:00 or 08:00:00Z), which successfully maps to Spanner STRING columns.

3. BYTEA, BIT, and VARBIT (Binary Data)
- Previous Behavior: Because Datastream converts Postgres binary columns to Base64 JSON strings, the pipeline previously failed to process them. It globally assumed all binary strings were Hex-encoded (which is only true for MySQL/Oracle), leading to data corruption and parsing errors when attempting to Hex-decode the Base64 values.
- New Behavior: The pipeline now checks the event's _metadata_source_type. If the source is postgresql, it safely uses a Base64 decoder to convert the string back to a Spanner BYTES array, preserving the exact binary data.

**Additional Changes**
- Expanded IT Coverage: Added integration tests for missing numeric and serial type mappings to Spanner (e.g., smallint_to_float64, serial8_to_numeric, int2_to_numeric).
- Test Maintainability: Replaced opaque, hardcoded Base64 values in ITs with explicit string-to-byte conversions (ByteArray.copyFrom("val").toBase64()), making it much easier to read the intended underlying values.

### CODE REVIEW

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java [hunk 1]
```diff
@@ -650,6 +654,53 @@ static void handleDatastreamRecordType(
                   .withZoneSameInstant(ZoneId.of("UTC"))
                   .format(DEFAULT_TIMESTAMP_WITH_TZ_FORMATTER));
           break;
+        case "timeTz":
+          Long timeTzNanos =
+              Long.valueOf(element.get("time").toString()) * TimeUnit.MICROSECONDS.toNanos(1);
+          int offsetSeconds = Integer.valueOf(element.get("offset").toString()) / 1000;
```
[the remark this answers is not in the document]
Comment (sm745052) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java]: @jsuhani-2026 can you look at this ?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 2]
```diff
@@ -618,15 +646,23 @@ private Map<String, List<Map<String, Object>>> getExpectedData() {
...
     result.put(
         "uuid_pk",
         createUuidPkRows(
             "a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11", "a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a12"));
     result.put("uuid_to_bytes", createRows("oO68mZwLTvi7bWu5vTgKEQ==", "NULL"));
-    result.put("varbit", createRows("wA==", "NULL"));
-    result.put("varbit_to_string", createRows("wA==", "NULL"));
+    result.put("uuid_to_string", createRows("a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11", "NULL"));
+    result.put("varbit", createRows("MTEwMA==", "NULL"));
+    result.put("varbit_to_string", createRows("31313030", "NULL"));
```
Comment (sm745052) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: why checked this ? should it not be binary string ?

Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: Datastream currently sends PostgreSQL VARBIT data as an Avro BYTES type containing the ASCII characters. Our pipeline automatically hex-encodes all BYTES mapped to Spanner STRING, which is why "1100" turns into its hex equivalent "31313030".

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java [hunk 3]
```diff
@@ -650,6 +654,53 @@ static void handleDatastreamRecordType(
...
+          if (!intervalDuration.isZero()) {
+            if (result.length() == 0) {
+              result.append(intervalDuration.toString());
+            } else {
+              // Remove the 'P' from Duration
+              result.append(intervalDuration.toString().substring(1));
+            }
+          }
+          jsonObject.put(fieldName, result.toString());
+          break;
```
[the remark this answers is not in the document]
Comment (jsuhani-2026) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java:703]: I tested this locally, and java.time.Duration.toString() does not use a leading minus sign (e.g. -PT2H...). Instead, it applies the minus sign to the specific components (e.g. PT-2H-15M). Because of this, substring(1) correctly extracts T-2H-15M and concatenates perfectly to form a valid ISO-8601 string. The current implementation works correctly for negative durations, so no changes are needed here.

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsuhani-2026) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java]: Done

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerITBase.java [hunk 4]
```diff
@@ -354,7 +354,7 @@ protected LaunchInfo launchDataflowJob(
       params.put("dlqGcsPubSubSubscription", dlqSubscription.toString());
     }
     params.put("inputFileFormat", "avro");
-    params.put("workerMachineType", "n2-standard-4");
+    params.put("workerMachineType", "e2-standard-4");
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerITBase.java]: why have we changed this?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 5]
```diff
@@ -297,17 +291,14 @@ private void validateResult(
...
-            "bytea",
-            "json",
-            "json_to_string",
-            "macaddr",
-            "macaddr8",
+            "bytea_to_string",
+            "time",
+            "time_with_time_zone",
+            "time_without_time_zone",
+            "timetz",
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:295]: why are time types ignored?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/source/postgresql/PostgresChangeEventContextTest.java [hunk 6]
```diff
@@ -97,6 +100,9 @@ public void canGenerateShadowTableMutationForBackfillEvent() throws Exception {
     changeEvent.put(PostgresqlDsToSpSourceConnector.POSTGRES_LSN_KEY, JSONObject.NULL);
     changeEvent.put(
         DatastreamConstants.EVENT_SOURCE_TYPE_KEY, SourceConstants.POSTGRES_SOURCE_TYPE);
+    changeEvent.put(
+        "bytes_field",
+        java.util.Base64.getEncoder().encodeToString(new byte[] {120, 53, 56, 48, 48}));
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/source/postgresql/PostgresChangeEventContextTest.java]: use imports rather than fully classified paths

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 7]
```diff
@@ -533,38 +547,42 @@ private Map<String, List<Map<String, Object>>> getExpectedData() {
     result.put("inet", createRows("192.168.1.0/24", "NULL"));
     result.put("int", createRows("-2147483648", "2147483647", "1", "NULL"));
     result.put("int_to_string", createRows("-2147483648", "2147483647", "1", "NULL"));
+    result.put("int_to_numeric", createRows("-2147483648", "2147483647", "1", "NULL"));
+    result.put("int_to_float64", createRows("-2.147483648E9", "2.147483647E9", "1.0", "NULL"));
     result.put("integer", createRows("-2147483648", "2147483647", "2", "NULL"));
     result.put("integer_to_string", createRows("-2147483648", "2147483647", "2", "NULL"));
+    result.put("integer_to_numeric", createRows("-2147483648", "2147483647", "2", "NULL"));
+    result.put("integer_to_float64", createRows("-2.147483648E9", "2.147483647E9", "2.0", "NULL"));
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:535]: will int to float64 not fail for boundary values, like we don't support it for bulk?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 8]
```diff
@@ -473,26 +466,46 @@ private Map<String, List<Map<String, Object>>> getExpectedData() {
...
+        "bit_varying_to_string",
+        createRows(Hex.encodeHexString("0101".getBytes(StandardCharsets.UTF_8)), "NULL"));
     result.put("bool", createRows("false", "true", "NULL"));
     result.put("bool_to_string", createRows("false", "true", "NULL"));
     result.put("boolean", createRows("false", "true", "NULL"));
     result.put("boolean_to_string", createRows("false", "true", "NULL"));
-    result.put("bytea", createRows("YWJj", "NULL"));
-    result.put("bytea_to_string", createRows("YWJj", "NULL"));
+    result.put("bytea", createRows(ByteArray.copyFrom("abc").toBase64(), "NULL"));
+    result.put(
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: Did we not decide to not support byte to string?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 9]
```diff
@@ -592,17 +613,40 @@ private Map<String, List<Map<String, Object>>> getExpectedData() {
...
+    result.put("smallint_to_numeric", createRows("-32768", "32767", "10", "NULL"));
+    result.put("smallint_to_float32", createRows("-32768.0", "32767.0", "10.0", "NULL"));
+    result.put("smallint_to_float64", createRows("-32768.0", "32767.0", "10.0", "NULL"));
     result.put("smallserial", createRows("-32768", "32767", "11"));
     result.put("smallserial_to_string", createRows("-32768", "32767", "11"));
+    result.put("smallserial_to_numeric", createRows("-32768", "32767", "11"));
+    result.put("smallserial_to_float32", createRows("-32768.0", "32767.0", "11.0"));
+    result.put("smallserial_to_float64", createRows("-32768.0", "32767.0", "11.0"));
     result.put("text", createRows("testing text", "NULL"));
+    // Datastream incorrectly wraps 24:00:00 to 0 microseconds during extraction.
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:622]: We should add this comment in the DatstreamIO code and test for time type other values than 24:00:00 instead of ignoring, it we can add a commented out test for 24:00:00

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java [hunk 10]
```diff
@@ -618,15 +662,25 @@ private Map<String, List<Map<String, Object>>> getExpectedData() {
         "timestamp_with_timezone_to_string",
         createRows("1970-02-02T18:05:06.123456Z", "1970-02-03T05:05:06.123456Z", "NULL"));
     result.put("timestamp_without_time_zone", createRows("1970-01-02T03:04:05.123456Z", "NULL"));
+    result.put(
+        "timestamp_without_time_zone_to_timestamp",
+        createRows("1970-01-02T03:04:05.123456000Z", "NULL"));
+    result.put("tsquery", createRows("'fat' & 'rat'", "NULL"));
+    result.put("tsvector", createRows("'a' 'cat' 'fat' 'mat' 'on' 'sat'", "NULL"));
+    result.put("txid_snapshot", createRows("10:20:10,14,15", "NULL"));
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:650]: are we supporting these in bulk?

File: v2/datastream-common/src/test/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJsonTest.java [hunk 11]
```diff
@@ -316,6 +316,68 @@ public void testIntervalNano() throws JsonProcessingException {
...
+        "zero_interval", generateIntervalSchema(), generateIntervalRecord(0, 0, 0L), objectNode);
+
+    /* Only months */
+    UnifiedTypesFormatter.handleDatastreamRecordType(
+        "only_months", generateIntervalSchema(), generateIntervalRecord(12, 0, 0L), objectNode);
+
+    /* Only time (5 hours + 123456 micros = 5H0.123456S) */
+    UnifiedTypesFormatter.handleDatastreamRecordType(
+        "only_time", generateIntervalSchema(), generateIntervalRecord(0, 5, 123456L), objectNode);
+
```
Comment (shreyakhajanchi) [inline v2/datastream-common/src/test/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJsonTest.java:338]: Add a test for negative interval as well

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java [hunk 12]
```diff
@@ -650,6 +653,54 @@ static void handleDatastreamRecordType(
                   .withZoneSameInstant(ZoneId.of("UTC"))
                   .format(DEFAULT_TIMESTAMP_WITH_TZ_FORMATTER));
           break;
+        case "timeTz":
```
Comment (shreyakhajanchi) [inline v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java:656]: why are we not follwoing the logic in https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/f263d5aee21c01f442d0afa319ef8d2a1b0167eb/v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java#L635

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java [hunk 13]
```diff
@@ -650,6 +653,54 @@ static void handleDatastreamRecordType(
...
+          if (timeTzNanos == 86400000000000L) {
+            jsonObject.put(fieldName, "24:00:00" + timeTzOffset.toString());
+            break;
+          }
+
+          LocalTime localTime = LocalTime.ofNanoOfDay(timeTzNanos);
+          OffsetTime offsetTime = OffsetTime.of(localTime, timeTzOffset);
+          jsonObject.put(fieldName, offsetTime.format(DateTimeFormatter.ISO_OFFSET_TIME));
+          break;
+        case "interval":
```
Comment (shreyakhajanchi) [inline v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java:673]: can we not reuse the code as in intervalNano?

**Review by shreyakhajanchi — COMMENTED** [review]
Can you please check that the type behaviour is consistent between bulk and live migration templates?

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerITBase.java (hunk 4 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/DataStreamToSpannerITBase.java]: The ITs were failing because of the compute engine stockout issue so by changing this the ITs are indeed passing.
Darshan is working on fixing this, can revert it when that is done.

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 5 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:295]: To avoid the failing test, because a value of 24hrs is converted to 0hrs by datastream.

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 7 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:535]: No int can have precision upto 10 digits and float64 supports precision upto 15-17 digits. 
We support it for bulk too

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 8 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: Yes, and they are not supported... they are added in the ignoredTypeMappings

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 10 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:650]: No they are not supported in bulk

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java (hunk 12 above)
Comment (jsuhani-2026) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java:656]: Because timestamptz has the date component so converting it to UTC will not lead to any info loss
But timeTz only contains the time. If we normalize a time like 02:00:00+05:00 to UTC, it becomes 21:00:00Z and we completely lose the information that a day-wrap occurred.

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java (hunk 13 above)
Comment (jsuhani-2026) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJson.java:673]: We cannot reuse the intervalNano logic for interval because the Avro input structures are fundamentally different between the two types. In the input schema, an interval record only has 3 fields (months, hours, and micros), requiring us to manually calculate days by dividing hours by 24. In contrast, intervalNano expects 7 explicit fields (years, months, days, hours, minutes, seconds, and nanos). Passing an interval record would result in silent data loss because the logic would fail to find the expected time fields and completely ignore the micros field.

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/source/postgresql/PostgresChangeEventContextTest.java (hunk 6 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/source/postgresql/PostgresChangeEventContextTest.java]: Done

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 9 above)
Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:622]: Done

File: v2/datastream-common/src/test/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJsonTest.java (hunk 11 above)
Comment (jsuhani-2026) [reply v2/datastream-common/src/test/java/com/google/cloud/teleport/v2/datastream/transforms/FormatDatastreamRecordToJsonTest.java:338]: Done

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 8 above)
Comment (shreyakhajanchi) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: let's remove it

Comment (jsuhani-2026) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java]: Done

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java (hunk 5 above)
Comment (shreyakhajanchi) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/PostgreSQLDatastreamToSpannerDataTypesIT.java:295]: Created a separate ticket to track this: b/543730685

**Review by shreyakhajanchi — APPROVED** [review]
